### PR TITLE
WdlObject implementation

### PR DESCRIPTION
**Pull request by Horneth**
Also coerce WdlMap <-> WdlObject whenever possible

### CODE REVIEW

File: src/main/scala/cromwell/binding/AstTools.scala [hunk 1]
```diff
@@ -76,6 +77,13 @@ object AstTools {
             k -> v
           }).toMap
           WdlMap(mapType, elements)
+        case a: Ast if a.getName == "ObjectLiteral" && wdlType == WdlObjectType =>
+          val elements = a.getAttribute("map").asInstanceOf[AstList].asScala.toVector.map({ kvnode =>
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/AstTools.scala]: Map over a non-anonymous function. I realize it's done this way right above you but I almost certainly made the same comment when that code was written too

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
+import scala.util.{Failure, Success}
 
 case object WdlObjectType extends WdlType {
   val toWdlString: String = "Object"
 
   override protected def coercion = {
     case o: WdlObject => o
+    case m: WdlMap if m.wdlType.keyType == WdlStringType && WdlStringType.isCoerceableFrom(m.wdlType.valueType) =>
+      val coercedMap = m.value map {
+        case (k,v) => k.asInstanceOf[WdlString].value -> WdlStringType.coerceRawValue(v).map(_.asInstanceOf[WdlString])
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: space after ,

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala [hunk 3]
```diff
@@ -1,16 +1,30 @@
...
+      } collect {
+        case (k, Success(v)) => k -> v
+        case (_, v: Failure[_]) => throw new UnsupportedOperationException(s"Unable to convert value to String")
+      }
+      WdlObject(coercedMap)
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: \n

File: src/main/scala/cromwell/binding/values/WdlObject.scala [hunk 4]
```diff
@@ -1,14 +1,41 @@
...
 trait WdlObjectLike {
   def value: Map[String, WdlValue]
 }
 
+object WdlObject {
+
+  def coerceObject(m: Map[String, String]): WdlObject = {
+    val coerced = WdlMap.coerceMap(m, WdlMapType(WdlStringType, WdlStringType)).value map {
+      case (k,v) => k.asInstanceOf[WdlString].value -> v
+    }
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: \n

File: src/main/scala/cromwell/binding/values/WdlObject.scala [hunk 5]
```diff
@@ -1,14 +1,41 @@
...
+  }
+
+}
+
 case class WdlObject(value: Map[String, WdlValue]) extends WdlValue with WdlObjectLike {
   val wdlType = WdlObjectType
+
+  override def toWdlString: String =
+    "object {" + value.map {case (k,v) => s"$k: ${v.toWdlString}"}.mkString(", ") + "}"
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: space after the ,

File: src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala [hunk 6]
```diff
@@ -11,7 +11,11 @@ object WdlValueJsonFormatter extends DefaultJsonProtocol {
       case f: WdlFloat => JsNumber(f.value)
       case b: WdlBoolean => JsBoolean(b.value)
       case f: WdlFile => JsString(f.value)
-      case o: WdlObject => JsObject()
+      case o: WdlObject => new JsObject(
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: all on one line

File: src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala [hunk 7]
```diff
@@ -11,7 +11,11 @@ object WdlValueJsonFormatter extends DefaultJsonProtocol {
...
+        Map("object" ->
+          JsObject(o.value map {case(k,v) => k -> write(v)})
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: space after the ,

**Comment by cjllanwarne** [comment]
Can confirm the Tyburn test "wdlobjects" works under this branch. Hurrah!

File: src/main/scala/cromwell/binding/values/WdlObject.scala (hunk 5 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: https://groups.google.com/forum/#!topic/workflow-description-language/h5W7MVkoeGk for discussion about the object literal syntax.  `object {...}` is a bit hacked into the parser (from back when I was writing PyWDL and I just needed something quick)

I don't think it needs to be addressed in this PR though.

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala [hunk 8]
```diff
@@ -29,12 +29,19 @@ class LocalEngineFunctionsWithoutCallContext extends WdlStandardLibraryFunctions
...
   }
 
+  override protected def read_object(params: Seq[Try[WdlValue]]): Try[WdlObject] = {
+    for {
+      singleArgument <- extractSingleArgument(params)
+      contents <- Success(fileContentsToString(singleArgument))
+      wdlObject <- WdlObject.fromTsv(contents)
+    } yield wdlObject
+  }
+
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: Could be nice to implement `write_object()` too.  If not part of this PR, maybe file a ticket for implementing `write_object`, `read_objects`, and `write_objects`?

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala [hunk 9]
```diff
@@ -1,16 +1,30 @@
...
+import cromwell.binding.values.{WdlCallOutputsObject, WdlMap, WdlObject, WdlString}
+
+import scala.util.{Failure, Success}
 
 case object WdlObjectType extends WdlType {
   val toWdlString: String = "Object"
 
   override protected def coercion = {
     case o: WdlObject => o
+    case m: WdlMap if m.wdlType.keyType == WdlStringType && WdlStringType.isCoerceableFrom(m.wdlType.valueType) =>
```
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: I'm not sure I understand why the coercion from `Map[String, String]` to `Object` exists.  Is this used anywhere or added because it might be useful?

Also, since this chunk is used more than once (also in `isCoerceableFrom`, perhaps extract it?

```scala
m.wdlType.keyType == WdlStringType && WdlStringType.isCoerceableFrom(m.wdlType.valueType)
```

I wonder if the key can be coercable to string like the value:

```scala
WdlStringType.isCoerceableFrom(m.wdlType.keyType) && WdlStringType.isCoerceableFrom(m.wdlType.valueType)
```

File: src/main/scala/cromwell/binding/AstTools.scala [hunk 10]
```diff
@@ -55,13 +55,14 @@ object AstTools {
       astNode match {
         case t: Terminal if t.getTerminalStr == "string" && wdlType == WdlStringType => WdlString(t.getSourceString)
         case t: Terminal if t.getTerminalStr == "string" && wdlType == WdlFileType => WdlFile(t.getSourceString)
+        case t: Terminal if t.getTerminalStr == "identifier" && wdlType == WdlStringType => WdlString(t.getSourceString)
```
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/AstTools.scala]: I'm not sure this is correct, I don't think identifiers should be converted to strings like this.  What problem did adding this line solve?

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala (hunk 8 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: It should be simple enough to add to this PR, although if for some reason it doesn't make it before the end of the day I'll file a ticket for it considering the "Tuesday code freeze".

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala (hunk 9 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: I thought it might be useful but it's not used anywhere as far as I can tell.

File: src/main/scala/cromwell/binding/AstTools.scala (hunk 10 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/AstTools.scala]: Yes it felt weird but it wouldn't work without it because when you access an object member say myObject.myAttribute, myObject is a terminal `identifier` which was never getting resolved because there is no `case` for it.
Now convert it to a string may not be the right thing to do although it does seem to work.

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala (hunk 9 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: I don't remember if this is covered elsewhere, but do we support coercing `JsObject` into `WdlObject`?

File: src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala (hunk 6 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: About this, I nested the actual json representation into an 
```json
{ "object": {
  "foo" : "bar",
    ...
  }
}
```
to make the difference with a map. Not sure this is a the best approach. Any idea ?

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: ooh I see what you did... any reason why it can't serialize to JSON just like a `WdlMap` serializes?

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala (hunk 9 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: I don't think we do, we probably should though since it looks like `JsObject` to `WdlMap` is supported

File: src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala (hunk 6 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: Not really, I was thinking when getting the output you would want to know if the result you get is a `WdlMap` or a `WdlObject` but maybe it doesn't matter actually.

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: Personally I'd lean toward keeping them both represented the same way.

File: src/main/scala/cromwell/binding/types/WdlObjectType.scala [hunk 11]
```diff
@@ -1,16 +1,30 @@
...
+      val coercedMap = m.value map {
+        case (k,v) => k.asInstanceOf[WdlString].value -> WdlStringType.coerceRawValue(v).map(_.asInstanceOf[WdlString])
+      } collect {
+        case (k, Success(v)) => k -> v
+        case (_, v: Failure[_]) => throw new UnsupportedOperationException(s"Unable to convert value to String")
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/types/WdlObjectType.scala]: maybe propagate the exception and/or its message?

File: src/main/scala/cromwell/binding/values/WdlObject.scala [hunk 12]
```diff
@@ -1,14 +1,41 @@
...
+  def coerceObject(m: Map[String, String]): WdlObject = {
+    val coerced = WdlMap.coerceMap(m, WdlMapType(WdlStringType, WdlStringType)).value map {
+      case (k,v) => k.asInstanceOf[WdlString].value -> v
+    }
+    WdlObject(coerced)
+  }
+
+  def fromTsv(tsv: String): Try[WdlObject] = {
+    FileUtil.parseTsv(tsv) match {
+      case Success(table) if table.isEmpty => Success(WdlObject(Map.empty[String, WdlValue]))
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: sorry, why is this ok?

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: That's a good point, I my guess is that it was probably copied/pasted from `WdlMap.fromTsv()`.  In this context I think you're right, this case should be an error

File: src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala (hunk 6 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlValueJsonFormatter.scala]: :+1:

File: src/main/scala/cromwell/binding/values/WdlObject.scala (hunk 12 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: I'm not sure I see why it's ok to parse an empty tsv file to a map but not to an object though

Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: Fair point.  I suppose I don't have a super compelling argument...  The way I see it, an empty map still conveys some meaning, whereas an empty object seems much more useless.  It's up to you, though

Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: Yeah I think I see what you mean, you could possibly try to iterate over an empty map for instance (if you don't know it's empty), whereas an empty object is completely useless.

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala [hunk 13]
```diff
@@ -3,10 +3,11 @@ package cromwell.engine.backend.local
 import java.io.File
 import java.nio.file.{Path, Paths}
 import cromwell.binding.expression.WdlStandardLibraryFunctions
-import cromwell.binding.types.{WdlArrayType, WdlFileType, WdlMapType, WdlStringType}
+import cromwell.binding.types._
 import cromwell.binding.values._
 import cromwell.util.FileUtil
 import cromwell.util.FileUtil.{EnhancedFile, EnhancedPath}
+import WdlArrayType._
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: use fully qualified path

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala [hunk 14]
```diff
@@ -96,9 +111,32 @@ class LocalEngineFunctions(cwd: Path, stdout: Path, stderr: Path) extends LocalE
...
+      singleArgument <- extractSingleArgument(params)
+      if singleArgument.wdlType.isAnArrayOf(WdlObjectType)
+      tsvSerialized <- singleArgument.asInstanceOf[WdlArray].tsvSerialize
+      file = writeContent("array", tsvSerialized)
+    } yield file
+  }
+
+  protected def writeContent(baseName: String, content: String) = {
+    val (path, writer) = FileUtil.tempFileAndWriter("array", cwd.toFile)
+    Try(writer.write(content))
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: Is this ignoring any error that might happen with `writer.write`?

File: src/main/scala/cromwell/binding/values/WdlObject.scala [hunk 15]
```diff
@@ -1,14 +1,85 @@
...
+    def attributesMatch(objects: Seq[WdlObject]): Boolean = {
+      val attributesSet = objects map { _.orderedAttributes }
+      val intersection = attributesSet reduce (_.intersect(_))
+      attributesSet forall { _ == intersection }
+    }
+
+    input match {
+      case Nil => Failure(new UnsupportedOperationException("Cannot write empty objects array."))
+      case objects =>
+        if (attributesMatch(objects)) {
```
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: perhaps `case objects if attributesMatch(objects) =>`?

**Comment by scottfrazer** [comment]
:+1: how about @cjllanwarne as second reviewer?

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala [hunk 16]
```diff
@@ -27,3 +27,14 @@ case class WdlArrayType(memberType: WdlType) extends WdlType {
     case _ => false
   }
 }
+
+object WdlArrayType {
+
+  implicit class WdlArrayEnhanced(wdlType: WdlType) extends WdlType {
+
+    override protected def coercion: PartialFunction[Any, WdlValue] = wdlType.coercion
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: My scala's not great on implicits... I don't see how this enhances the basic coercion behaviour?

File: src/main/scala/cromwell/binding/values/WdlArray.scala [hunk 17]
```diff
@@ -25,6 +25,7 @@ case class WdlArray(wdlType: WdlArrayType, value: Seq[WdlValue]) extends WdlValu
   def tsvSerialize: Try[String] = {
     wdlType.memberType match {
       case t: WdlPrimitiveType => Success(value.map(_.valueString).mkString("\n"))
+      case WdlObjectType => WdlObject.tsvSerializeArray(value map { _.asInstanceOf[WdlObject] })
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/binding/values/WdlArray.scala]: Is this the point where we should check all Objects have the same fields?

File: src/main/scala/cromwell/binding/values/WdlObject.scala [hunk 18]
```diff
@@ -1,14 +1,84 @@
...
+      case Success(table) if table.isEmpty => Failure(new UnsupportedOperationException("TSV file was empty or could not be parsed."))
+      case Success(table) if table.length < 2 => Failure(new UnsupportedOperationException("TSV must be 2 rows (or more) to convert to an Object (Array[Object])"))
+      case Success(table) => Try {
+        table.tail map { line => coerceObject((table.head zip line).toMap) }
+      }
+      case Failure(e) => Failure(e)
+    }
+  }
+
+  def tsvSerializeArray(input: Seq[WdlObject]) = {
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: So this may not be a problem, and it may be outside the scope of this ticket, but generating entire files in memory is generally a slightly questionable idea... especially if we're talking about several GB/TB/PB of data...

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala [hunk 19]
```diff
@@ -96,9 +112,37 @@ class LocalEngineFunctions(cwd: Path, stdout: Path, stderr: Path) extends LocalE
...
+    for {
+      singleArgument <- extractSingleArgument(params)
+      if singleArgument.wdlType.isAnArrayOf(WdlObjectType)
+      tsvSerialized <- singleArgument.asInstanceOf[WdlArray].tsvSerialize
+      file <- writeContent("array", tsvSerialized)
+    } yield file
+  }
+
+  protected def writeContent(baseName: String, content: String) = {
+    val (path, writer) = FileUtil.tempFileAndWriter("array", cwd.toFile)
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: Is this going to clash if the same task writes two arrays?

e.g. task x {
 ...
 ...
 output {
  File a = write_array(blah)
  File b = write_array(blah2)
 }
}

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: I think `tempFileAndWriter` adds a random string onto "array"

File: src/main/scala/cromwell/binding/values/WdlArray.scala (hunk 17 above)
Comment (scottfrazer) [inline src/main/scala/cromwell/binding/values/WdlArray.scala]: One could check at WdlArray construction that all objects have the same fields, but it only *really* matters if they have the same fields if you try to serialize it.  There's no reason why you can't have heterogeneous objects in an array if you never try to serialize the array into a TSV file.

File: src/test/scala/cromwell/binding/types/WdlObjectTypeSpec.scala [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+  it should "NOT successfully coerce a NON coerceable map into a WdlObject (2)" in {
+    WdlObjectType.coerceRawValue(nonCoerceableMap2) match {
+      case Success(v) => fail("should not have succeeded")
+      case Failure(f) => // expected
+    }
+  }
+
+  it should "NOT successfully convert WDL source code to WdlMap if passed a bogus AST" in {
+    try {
+      WdlObjectType.fromWdlString("workflow wf{}")
```
Comment (cjllanwarne) [inline src/test/scala/cromwell/binding/types/WdlObjectTypeSpec.scala]: Is there a reason fromWdlString doesn't return a Try?

**Comment by cjllanwarne** [comment]
:+1:

File: src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala (hunk 19 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/local/LocalEngineFunctions.scala]: Yes it does, it'll be something like `arrayfds65145gfd3.tmp`

File: src/test/scala/cromwell/binding/types/WdlObjectTypeSpec.scala (hunk 20 above)
Comment (Horneth) [inline src/test/scala/cromwell/binding/types/WdlObjectTypeSpec.scala]: Hmm I guess it could, it's only used in one place though (+ in the tests) to convert a symbol DB entry to a WdlValue, which should never fail.
I think this could be included as part of a larger refactoring/robustifying, I'll add a TODO.

File: src/main/scala/cromwell/binding/types/WdlArrayType.scala (hunk 16 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/types/WdlArrayType.scala]: It doesn't, but because it extends WdlType it needs to implement `coercion` and `toWdlString`, so we reuse the already existing implementation.
The added value is the `isAnArrayOf` method a few lines below which removes some boilerplate.

File: src/main/scala/cromwell/binding/values/WdlArray.scala (hunk 17 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlArray.scala]: This is done in the `tsvSerializeArray` function (there is a `attributesMatch` function)

File: src/main/scala/cromwell/binding/values/WdlObject.scala (hunk 18 above)
Comment (Horneth) [inline src/main/scala/cromwell/binding/values/WdlObject.scala]: I agree, some kind of stream would be better to avoid flushing the whole string in memory just to write it to a file a few method calls later. I'll add a FIXME.
